### PR TITLE
Field id support

### PR DESCRIPTION
**Pull request by TylerBrinkley**
😄

### CODE REVIEW

File: src/EncompassRest.Tests/LoanTests.cs [hunk 1]
```diff
@@ -357,5 +357,171 @@ public void Loan_ExtensionData_Serialization()
...
+            var loan = new Loan();
+            var value = "Joe";
+            loan.Fields["4000#2"] = value;
+            Assert.AreEqual(value, (string)loan.Fields["4000#2"]);
+            Assert.AreEqual($@"{{""applications"":[{{""applicationIndex"":1,""borrower"":{{""firstName"":""{value}""}}}}]}}", loan.ToJson());
+            loan.Fields["4000#2"] = null;
+            Assert.IsNull(loan.Fields["4000#2"]);
+            Assert.AreEqual($@"{{""applications"":[{{""applicationIndex"":1,""borrower"":{{""firstName"":null}}}}]}}", loan.ToJson());
+        }
     }
```
Comment (hfcjweinstock) [inline src/EncompassRest.Tests/LoanTests.cs]: Add Loan_FieldsInvalidField and LoanFieldsInvalidCustomField
Both should throw ArgumentException

File: src/EncompassRest/Loans/LoanFieldsCustomizations.cs [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+                        {
+                            return customField.DateValue;
+                        }
+                        if (customField.NumericValue.HasValue)
+                        {
+                            return customField.NumericValue;
+                        }
+                        return customField.StringValue;
+                    }
+                    return null;
```
Comment (hfcjweinstock) [inline src/EncompassRest/Loans/LoanFieldsCustomizations.cs]: throw ArgumentException

File: src/EncompassRest/Loans/LoanFieldsCustomizations.cs [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+            {
+                Preconditions.NotNullOrEmpty(fieldId, nameof(fieldId));
+
+                fieldId = fieldId.ToUpper();
+                if (fieldId.StartsWith("CX.", StringComparison.Ordinal))
+                {
+                    var customField = _loan.CustomFields.FirstOrDefault(f => string.Equals(fieldId, f.FieldName, StringComparison.OrdinalIgnoreCase));
+                    if (customField == null)
+                    {
+                        customField = new CustomField { FieldName = fieldId };
```
Comment (hfcjweinstock) [inline src/EncompassRest/Loans/LoanFieldsCustomizations.cs]: this should check if the field is available in the schema.  if not, throw argument exception.

Comment (TylerBrinkley) [reply src/EncompassRest/Loans/LoanFieldsCustomizations.cs]: What if they're just updating a field on a loan that hasn't been retrieved? We would have no idea what custom fields are available.

Comment (hfcjweinstock) [reply src/EncompassRest/Loans/LoanFieldsCustomizations.cs]: it would still show up in the schema call.  possibly use getSchemaField ?

Comment (TylerBrinkley) [reply src/EncompassRest/Loans/LoanFieldsCustomizations.cs]: You cannot retrieve custom fields. Besides, that's an asynchronous call.

Comment (hfcjweinstock) [reply src/EncompassRest/Loans/LoanFieldsCustomizations.cs]: add boolean option for createIfNotExist?  there has to be some "validation" that what you are writing is valid.

**Comment by hfcjweinstock** [comment]
does this handle field 2181? or any complex mapped fields?

**Comment by TylerBrinkley** [comment]
Not currently, it only supports fields that are direct descendants of the Loan object, i.e. not a member of a list, except for borrower pair fields. I could probably add support for more complex fields with the addition of an indexer syntax.

**Comment by TylerBrinkley** [comment]
I was looking at the `FieldLockData` and it has the perfect syntax for us.

```js
{
    "fieldLockData": [
        {
            "modelPath": "Loan.ClosingCost.DisclosedSalesPrice"
        },
        {
            "modelPath": "Loan.Contacts[(ContactType == 'BROKER_LENDER')].Name"
        },
        {
            "modelPath": "Loan.RateLock.BuySideAdjustments[(PriceAdjustmentType == 'BasePrice' && AdjustmentType == 'Adjustment')][10].Rate"
        }
    ]
}
```

**Comment by hfcjweinstock** [comment]
been rolling this around for a while.  do we want to spin this to a feature branch until EM publishes the custom field management api next month?

**Comment by TylerBrinkley** [comment]
I don't think we should do custom field id validation as that requires async api calls which we shouldn't do from an indexer.

**Comment by hfcjweinstock** [comment]
we could call the index listing once and store it on a session basis.

**Comment by hfcjweinstock** [comment]
what does api do if you try to write to an invalid custom field?

**Comment by TylerBrinkley** [comment]
It probably silently fails. Perhaps we could add an option on loan update or create that would first validate the custom fields included.

**Comment by TylerBrinkley** [comment]
Only thing I'd like to do more with this is to add support for field patterns.

**Comment by TylerBrinkley** [comment]
@hfcjweinstock I believe this is ready to review. By the way, I tested invalid custom fields and they do return a 400 Bad request response so we do not need to do validation on the client side.

```js
{
    "summary": "Bad Request",
    "details": "Invalid CustomField Id ='', FieldName = 'CX.ABC'\r\n"
}
```

**Pull request merged**